### ISSUE TITLE
Controller Issues: Xbox One

### ISSUE THREAD

**Issue by bladehorton** [TeamMidnightDust/MidnightControls#79]
**Describe the bug**
Controller input is not detected by Minecraft

**To Reproduce**
1. Download Midnight Controls
2. Start Minecraft
3. Connect Xbox One controller via USB or bluetooth
4. See controller detected message
5. Change settings to suit you
6. Have controller not work

**Expected behavior**
correctly functioning controller

**Desktop (please complete the following information):**
 - OS: Linux Mint
 - Minecraft 1.19
 - Fabric 0.11.0
 - Mods modmenu v1.7.15
 - Version [e.g. 1.4.1
 - Branch Release

**Additional context**
No input is detcted despite settings being correct and controller is shown to be detected my Minecraft. Anytime I try and press the re-bind button it immediately binds to LT+RT

**Comment by bladehorton** [comment]
(NOTE: Pro controllers don't work either)

**Comment by Motschen** [comment]
Does the controller name appear orange in the settings?

**Comment by doinkers** [comment]
Can reproduce on Arch Linux.
The Xbox Series controller works, but the Xbox One controller does not.

> Does the controller name appear orange in the settings?

It does. 

The Linux gamepad-tool app detects all of the button presses just fine, but the mod does nothing when buttons are pushed. 

Let me know if you need any logs or controller mappings uploading, I'd be happy to do so!

**Comment by Motschen** [comment]
Could you try if it also happens on 1.3.0?

**Comment by Motschen** [comment]
The orange text indicates missing mappings, try creating one using gamepad-tool (https://generalarcade.com/gamepadtool/) and paste it into the Custom Mappings section.
If that works, you might as well want to contribute the mapping to https://github.com/gabomdq/SDL_GameControllerDB so other people using the same Controller don't have to do that themselves.

**Comment by doinkers** [comment]
> The orange text indicates missing mappings, try creating one using gamepad-tool (https://generalarcade.com/gamepadtool/) and paste it into the Custom Mappings section. If that works, you might as well want to contribute the mapping to https://github.com/gabomdq/SDL_GameControllerDB so other people using the same Controller don't have to do that themselves.

I tried that, sorry I didn't make it clear in my reply. The gamepad-tool detects all the buttons correctly, but even after pasting the mappings into the Custom Mappings, it doesn't work.

**Comment by bladehorton** [comment]
> Does the controller name appear orange in the settings?

Yes it does. Both the Xbox and Pro controller show up as orange. I'll add this it a gen 2 controller that came with a One S, not a gen 1 or a Series X controller

> Can reproduce on Arch Linux. The Xbox Series controller works, but the Xbox One controller does not.
Seems a bit weird. I even installed all available drivers for 360 and One controllers alike, none changed anything

> The Linux gamepad-tool app detects all of the button presses just fine, but the mod does nothing when buttons are pushed.

Pretty much sums up the whole thing. What I don't get is why it does this with Switch controllers. I could understand DS4s, they are a bit weird, and Switch kinda makes sense, as you need a driver for it, bus Xbox controllers can technically work out of the box

**Comment by bladehorton** [comment]
> The orange text indicates missing mappings, try creating one using gamepad-tool (https://generalarcade.com/gamepadtool/) and paste it into the Custom Mappings section.

Also doesn't work. Plus the "missing mappings" was just for other mods I had installed at the time, but it still shows up as orange without them. Maybe missing the mod menu mappings

**Comment by Motschen** [comment]
With "missing mappings" I mean missing controller mappings

**Comment by Motschen** [comment]
Could you try if it also happens on MidnightControls 1.3.0?

**Comment by doinkers** [comment]
> Could you try if it also happens on MidnightControls 1.3.0?

It doesn't seem to work in 1.3.0 either. I used gamepad-tool again, still to no avail.

`050000005e040000200b000011050000,Xbox One S Controller,a:b0,b:b1,x:b3,y:b4,back:b10,guide:b12,start:b11,leftstick:b13,rightstick:b14,leftshoulder:b6,rightshoulder:b7,dpup:h0.1,dpdown:h0.0,dpleft:h0.8,dpright:h0.2,leftx:a0,lefty:a1,rightx:a2,righty:a3,lefttrigger:a5,righttrigger:a4,platform:Linux,`

**Comment by Motschen** [comment]
Weird. Have you tried using the `xone` driver? It aims to improve Xbox controller support on Linux. I can't replicate the issue, since the only Xbox controller I own is an Afterglow Xbox One controller and it works perfectly for me both on Fedora and Arch Linux.

**Comment by doinkers** [comment]
> Weird. Have you tried using the `xone` driver? It aims to improve Xbox controller support on Linux. I can't replicate the issue, since the only Xbox controller I own is an Afterglow Xbox One controller and it works perfectly for me both on Fedora and Arch Linux.

xone doesn't seem to support Bluetooth, but I'll try it wired just to test it out.

**Comment by doinkers** [comment]
Interesting. After installing the xone driver, the controller works when its plugged in. Even more interestingly, it now works properly over Bluetooth *after* unplugging the cable.
![image](https://user-images.githubusercontent.com/105917902/181574775-2940b669-970e-44d0-97c4-62d0b5a8655f.png)


So, a workaround would be to install the xone driver, plug it in, then connect the controller again over Bluetooth.

**Comment by bendavis78** [comment]
When I first connected the Xbox controller it worked fine, did not appear orange. Today when I tried connecting again it turned orange and no longer works.

I'm on SteamDeck, using MidnightControls 1.5.0+1.19

**Comment by Motschen** [comment]
> When I first connected the Xbox controller it worked fine, did not appear orange. Today when I tried connecting again it turned orange and no longer works.
> 
> I'm on SteamDeck, using MidnightControls 1.5.0+1.19

Yeah, the orange text indicates missing mappings. For some reason the 'xone' driver changes the controllers GUID it seems, so the mappings will no longer work correctly.
Try creating a custom mapping using gamepad-tool (https://generalarcade.com/gamepadtool/) and paste it into the Custom Mappings section.
If that works, you might as well want to contribute the mapping to https://github.com/gabomdq/SDL_GameControllerDB so other people using the same Controller with xone don't have to do that themselves.

**Comment by tr1p0p** [comment]
I got exactly the same issues. Using linux with xpadneo drivers. Controller in orange (it's...the same that like a xbox 360 controller, really you need a layout for this decade old layout ??) and it's not recognize input.

**Comment by dhalucario** [comment]
Hey, I have the same issue. To my knowledge we should not even need drivers since the kernel comes with a driver for XBOX controllers and joycons trough evdev. I am using the latest Fedora version.
https://gamepad-tester.com/ recognizes my controller just fine too and I am on a wireless xbone controller

**Comment by dhalucario** [comment]
https://gamepad-tester.com/ recognizes my controller just fine too.

**Comment by Motschen** [comment]
> Hey, I have the same issue. To my knowledge we should not even need drivers since the kernel comes with a driver for XBOX controllers and joycons trough evdev. I am using the latest Fedora version. https://gamepad-tester.com/ recognizes my controller just fine too and I am on a wireless xbone controller

Can you tell me if the controller also works in this program: https://generalarcade.com/gamepadtool/ ?

**Comment by dhalucario** [comment]
> > Hey, I have the same issue. To my knowledge we should not even need drivers since the kernel comes with a driver for XBOX controllers and joycons trough evdev. I am using the latest Fedora version. https://gamepad-tester.com/ recognizes my controller just fine too and I am on a wireless xbone controller
> 
> Can you tell me if the controller also works in this program: https://generalarcade.com/gamepadtool/ ?

Yes, I pressed some buttons for show too:
![image](https://user-images.githubusercontent.com/10533763/205344269-2bdeebb2-fca5-4352-a8e6-a046e32830b6.png)

**Comment by DivineEssentia** [comment]
I have the same issue

**Comment by BadKiko** [comment]
Same

**Comment by aaronjamt** [comment]
I'm having the same issue on Ubuntu 23.04 with an Xbox One X controller, both wired and wireless, and both with and without the `xone` driver installed. I wanted to add that when I enabled the debug setting and looked at the logs, the controller says "Gamepad: false". Could that be part of the issue? Here's the relevent section:
```
[MidnightControls] Controller #0 name: "Microsoft X-Box One S pad"
 GUID: 030000005e040000ea0200000f050000
 Gamepad: false
```

**Comment by Motschen** [comment]
All issues related to this specific controller should now be reported in this issue.

**Comment by jwtaufer** [comment]
> Interesting. After installing the xone driver, the controller works when its plugged in. Even more interestingly, it now works properly over Bluetooth _after_ unplugging the cable. ![image](https://user-images.githubusercontent.com/105917902/181574775-2940b669-970e-44d0-97c4-62d0b5a8655f.png)
> 
> So, a workaround would be to install the xone driver, plug it in, then connect the controller again over Bluetooth.

Mine works plugged in, but this workaround doesn't work for me.

**Comment by EzyPzyAsh** [comment]
Was there ever a solution to the controller not doing anything? Mine is orange even after putting in the mappings. I tried installing xone-dkms but it didnt solve the issue. I am using it wired on Arch Linux

**Comment by Motschen** [comment]
> Was there ever a solution to the controller not doing anything? Mine is orange even after putting in the mappings. I tried installing xone-dkms but it didnt solve the issue. I am using it wired on Arch Linux

That's probably because the gamepad tool detects a different controller GUID than MidnightControls does (for some strange reason). My plan is to add a helper function for such cases in the next update, but for now you have to manually select the relevant controller in MidnightControls and then click on the 'Copy Current GUID' button below the mappings editor field.
Replace the GUID in the mapping string with that one and load the new mapping :)
